feat(DSCO): add optional onClose prop to Tag

drizco opened the pull request:
<!--
  A summary of the change, including any relevant background, motivation, and context.
  If relevant, include a description, screenshots, and/or video of the existing and new behavior.
-->
This updates the `Tags` component with an onClose handler. The onClose is specified in the tag props and is expected to be a state handler that will remove the tag from the `tagsList`.

## Accessibility
- The close button component that is rendered in the Tag has additional attributes to help make it accessible when an onClose handler is passed
- Since the close button rendered by the tag is also tabbable, to prevent double tabbing, the tabIndex is removed from the container attributes

## Recording of storybook

https://github.com/user-attachments/assets/be2db9f5-3a63-40eb-bb00-952deefab897


## Links
[figma](https://www.figma.com/design/Km9JCLNbayBlTrlnKenumq/New-PL-Program?node-id=3335-30674&t=OqNAp4JrHKk73FoJ-0)
<!--
  Links to relevant external resources; ie, specification documents, Jira tickets, related PRs, Honeybadger errors, etc.
-->

<!--
- spec: []()
- jira ticket: []()
-->

## Testing story

<!--
  Does your change include appropriate tests?
  If so, please describe how the tests included in this PR are sufficient.
  If not, please explain why this change does not need to be tested.
-->

<!-- Other aspects to consider. Delete any sections that are not relevant to your change. -->

## Deployment strategy

## Follow-up work

<!--
  List (ideally with Jira links) any clean-up or technical debt that will be addressed in future work.
-->

## Privacy

<!--
  1.	Does this change involve the collection, use, or sharing of new Personal Data?
  2.	Does this change involve a new or changed use or sharing of existing Personal Data?
-->

## Security

<!-- Link to Jira task(s) where sensitive security issues are discussed privately. -->

## Caching

## PR Checklist:

<!--
  The final step! Before you create your PR, double-check that everything is in order.
  Change [ ] to [X] during creation to check boxes.
-->

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

moshebaricdo commented on frontend/packages/component-library/src/tags/__tests__/Tags.test.tsx:
In Figma, I nested our existing close button component into the tag, would it simplify the changes here if we added that instead? Here's the storybook entry: https://code-dot-org.github.io/code-dot-org/component-library-storybook/?path=/docs/designsystem-closebutton--docs

drizco replied:
replaced the clickable icon with the CloseButton component 🙌

levadadenys commented on frontend/packages/component-library/src/tags/Tags.tsx:
```diff
@@ -41,16 +41,18 @@ const Tags: React.FunctionComponent<TagsProps> = ({
+    {tagsList.map(
+      ({tooltipId, label, tooltipContent, ariaLabel, icon, key}, i) => (
+        <Tag
+          key={key ?? tooltipId ?? label ?? i}
```
Can you please explain this change?

key should never be an index cause it may cause unexpected regression when order of the array changes. https://react.dev/learn/rendering-lists#keeping-list-items-in-order-with-key The fact that it may be working in storybook story doesn't mean one should rely on it.

drizco replied:
for sure! I added the key prop, as the tooltipId was being overloaded as the key (particularly since it was an optional prop). I kept the key prop optional, to preserve backwards compatibility with existing usages, but it really should be required. I can remove the index fallback, it felt better than nothing at the time. but I guess react will fallback to array indices anyway per that article 🤷

drizco replied:
removed label and index

levadadenys replied:
Oh, gotcha! Thanks for the context! Yeah, you're right, thank you for spotting this! Looks like initially tooltipId was a required prop, but then it changed in https://github.com/code-dot-org/code-dot-org/pull/63461/files#diff-13642ab24ded4eb8930beaf10c6b8b92972df32a25012c5344dd9c715da2bdc8 :( 

Looking into current usage of Tags I'm happy that at least most of the tags use tooltips right now 😄 
I think in this case adding `label` as a final resort fallback key option is totally valid as realistically every tag should have a unique label. (we might actually want to update the storybook to reflect that since it may actually misleading that right now there are couple +1 labels at the same time)

I was thinking if we should rename `key` prop to `tagId`, but honestly `key` also feels reasonable. So I'm just letting you know I had this thoughts and leave it for you to decide

levadadenys commented on frontend/packages/component-library/src/tags/_Tag.tsx:
```diff
@@ -10,15 +11,42 @@ type TagIconProps = {
-));
+const TagIcon: React.FC<TagIconProps> = memo(
+  ({iconName, iconStyle, title, onClick}) => {
+    const onClickProps = onClick
```
Might be worth using closeButton component here instead and just add onClose prop to tag/tags

drizco replied:
wouldn't having an `onClose` prop mean the dev can't choose an icon or icon style? @moshebaricdo could there be other `onClick` functionality on tags besides removal?

drizco replied:
after discussing with @moshebaricdo , <s>we think it might make sense to have multiple `tagType`s</s>, with the default being the current tag with optional icon. then a "closeable" tag with an `onClose` prop and using the closeButton component. in the future there are designs for a "link" tag, so this would start in that direction. I'll make some updates and let you know when it's ready for review again. thanks @levadadenys !

drizco replied:
> we think it might make sense to have multiple tagTypes

I realized this would not be great for backwards compatibility, and the presence of onClose, or tooltipContent were easy enough to use to determine how to render the tag

levadadenys replied:
> I realized this would not be great for backwards compatibility, and the presence of onClose, or tooltipContent were easy enough to use to determine how to render the tag

Why not? We can make `type` prop for Tags component and make it optional with default value of 'regular'/'simple', etc. Can't we?

levadadenys replied:
Though, it also comes down to the question if we're sure that we won't want a closable link type tags 😄 , in this case your approach might be better

drizco replied:
> Why not? We can make `type` prop for Tags component and make it optional with default value of 'regular'/'simple', etc. Can't we?

ah, I gave it another whirl with the `type` optional and I think it's better now, thanks! 🙌 

> Though, it also comes down to the question if we're sure that we won't want a closable link type tags 😄 , in this case your approach might be better

oof, I feel fairly confident @moshebaricdo would never design a ui element with nested clickable elements! 😅

moshebaricdo replied:
Yes, we def don't want that 😆

levadadenys commented on frontend/packages/component-library/src/tags/stories/Tags.story.tsx:
```diff
@@ -23,6 +24,46 @@ const SingleTemplate: StoryFn<TagsProps> = args => (
+const SingleTemplateWithOnClick: StoryFn<TagsProps> = args => {
+  const [tags, setTags] = useState(['AAA', 'BBB', 'CCC']);
+
+  const swapTags = (index: number) => {
```
Looks like onClick should be there just for onClose functionality support, so it worth checking specifically that functionality in the test.

Also, in case we'll want to add reorder functionality to Tags some point of time later, I think we would want to make it via drag/drop whole Tag, not onClick

drizco replied:
yeah, this was just a storybook example of what you can do with an `onClick` prop

levadadenys reviewed (commented):
I know that this is probably in the process of using `CloseButton` right now, so left rather general comments for now. Excited to see this done though! 😄 

One more general comment - we usually add a videos/screen recordings when the change/update is this big. Please add one once it will be ready. It makes it easier for designers and developers to check the result

levadadenys commented on frontend/packages/component-library/src/tags/_Tag.tsx:
```diff
@@ -32,38 +34,67 @@ export interface TagProps {
+    'aria-label': ariaLabel,
+    'aria-describedby': tooltipContent && tooltipId ? tooltipId : undefined,
+  };
+  if (onClose) {
```
e.g. if(type === 'closable') {

levadadenys commented on frontend/packages/component-library/src/tags/_Tag.tsx:
```diff
@@ -32,38 +34,67 @@ export interface TagProps {
+      </div>,
+    );
+  }
+  return tooltipWrapper(
```
e.g. (if type === 'simple')

(not sure how to name this type 😄 )

levadadenys commented on frontend/packages/component-library/src/tags/__tests__/Tags.test.tsx:
```diff
@@ -143,4 +143,84 @@ describe('Design System - Tags Component', () => {
+      });
+    });
+
+    it('has keyboard navigable tag icons if onClick is present', async () => {
```
super nit: might be worth using `onClose` instead of `onClick` as it may be misleading

drizco replied:
oops! will update 😅

levadadenys commented on frontend/packages/component-library/src/tags/__tests__/Tags.test.tsx:
```diff
@@ -143,4 +143,84 @@ describe('Design System - Tags Component', () => {
+      expect(onClick2).not.toHaveBeenCalled();
+    });
+
+    it('calls onClick when Enter or Space is pressed while tag has focus', async () => {
```
super nit: might be worth using onClose instead of onClick as it may be misleading

levadadenys commented on frontend/packages/component-library/src/tags/tags.module.scss:
```diff
@@ -32,9 +34,14 @@
       padding-inline-end: 8px;
     }
+
+    .icon_clickable {
```
nit: are we using it anywhere? Can't find it in the code, and looks like CloseButton should add cutrosr pointer by itself

also rather for future - for classNames like this we usually use camelCase , not underscore (for consistency)

drizco replied:
gah, this is old as well. will remove!

levadadenys commented on frontend/packages/component-library/src/tags/_Tag.tsx:
```diff
@@ -32,38 +34,67 @@ export interface TagProps {
    *  and the placement of the icon (left or right))*/
   icon?: TagIconProps;
+  /** Unique key */
+  key?: string | number;
```
super nit: should we remove number type from here to make it harder to pass index as a key? 😄

drizco replied:
lol, I'd definitely just do `index.toString()`

I think we should probably match how react types keys to avoid confusion

drizco replied:
updated the type to be `React.Key`, which is the same as `string | number` but makes it a lot clearer to users

levadadenys reviewed (commented):
Looks great!
Have some rather small comments, but this is awesome overall!

moshebaricdo reviewed (approved):
Pending any outstanding technical details, designs and behavior look good to me!

drizco commented:
> # 🖼️ Visual Comparison Report
> ⚠️⚠️⚠️ Detected eyes differences, see [report](https://eyes.applitools.com/app/test-results/00000251660574038738?accountId=mG4DoRfv0ntEgI3XIYBulrKxlZqrHtGwbP3V103ojvLFU110)!
> 
> Applitools Eyes is used to perform visual comparison testing against the `staging` baseline.
> 
> ## Remediation steps:
> 1. Open the [report](https://eyes.applitools.com/app/test-results/00000251660574038738?accountId=mG4DoRfv0ntEgI3XIYBulrKxlZqrHtGwbP3V103ojvLFU110)
> 2. Determine whether the differences are expected based on this PR's changes
>    a. **If expected**: Before merging this PR, accept the new baselines and re-run this action, it should pass.
>    b. **If not expected**: Push updates to this PR to correct the differences.

these changes are expected since I changed some tag labels and added the closable tag to the tag size groups

levadadenys commented on frontend/packages/component-library/src/tags/_Tag.tsx:
```diff
@@ -32,39 +33,81 @@ export interface TagProps {
+  onClose: () => void;
+}
+
+export type TagProps = DefaultTagProps | CloseTagProps;
```
🤩

levadadenys replied:
Though, super nit: Maybe `ClosableTagProps` instead of `CloseTagProps` 😄 ?

levadadenys commented on frontend/packages/component-library/src/tags/stories/Tags.story.tsx:
```diff
@@ -23,6 +24,41 @@ const SingleTemplate: StoryFn<TagsProps> = args => (
+    label,
+    tooltipId: label,
+    tooltipContent: label,
+    onClose: () => removeTag(i),
```
nit: we now need to add `type: 'closable'` here to make this work (checked locally)

```suggestion
    onClose: () => removeTag(i),
    type: 'closable',

drizco replied:
oof! thanks!

levadadenys commented on frontend/packages/component-library/src/tags/_Tag.tsx:
```diff
@@ -32,39 +33,81 @@ export interface TagProps {
+    const {onClose} = props as CloseTagProps;
+    return tooltipWrapper(
+      <div {...containerAttrs}>
+        <span>{label}</span>
```
(not blocking, can be done as a follow up pr/discussion)
How about making it like this?
```suggestion
        {/* eslint-disable-next-line jsx-a11y/no-noninteractive-tabindex */}
        <span tabIndex={0}>{label}</span>
```
I know it's not ideal, but it will make tooltip be keyboard and screen reader accessible and we won't be nesting focusable elements (cause with current solution only close button will be accessible)?

Also, as another option - maybe we can remove tooltips usage for closable tags?

Here's a video of how it will look like:

https://github.com/user-attachments/assets/dc64d778-9448-4e16-8922-5f3548a70f43



cc @moshebaricdo @hannahbergam @markabarrett

drizco replied:
yeah, I really wasn't sure what we should do here. I felt like having to tab twice to get through an element was bad, and the tooltip does still pop when the close button gets focus 🤷

moshebaricdo replied:
Hmm, agreed that this isn't super ideal, it's odd to see the tooltip appear both times, is it possible to have the tooltip appear when content/label is focused, but not for the close button? I don't want to get rid of tooltip support entirely for this tag type, maybe it becomes optional so we can decide depending on context.

drizco replied:
here's a video of the existing implementation to compare:


https://github.com/user-attachments/assets/fe9106d5-406d-4b6b-b5a9-985f3208bd7e

moshebaricdo replied:
Ok I dug through about 9 systems that all had a removable tag type, and it seems like they all take the approach of only making the close button focusable and not the label—and they all do not use tooltips for this case lol.

drizco replied:
the tooltip is definitely optional (only shows if you pass the correct props, tooltipContent and tooltipId)

as far as preventing it from showing when the close button has focus, I'd have to dig into the tooltip implementation, but I'm pretty sure it would be tricky. the tooltip is a wrapper for any component

levadadenys commented on frontend/packages/component-library/src/tags/_Tag.tsx:
```diff
@@ -32,39 +33,81 @@ export interface TagProps {
+        <CloseButton
+          onClick={onClose}
+          aria-label={`Close ${ariaLabel ?? label}`}
+          tabIndex={0}
```
nit: do we need tabIndex here? `CloseButton` is already a `button` html element

drizco replied:
ah, nope! good catch, this was probably a carryover of when I made the close icon clickable. removing

levadadenys reviewed (approved):
Looks great! Thank you very much for all the updates!
I left some final rather smaller comments and one comment that can also be done as a followup (tabIndex for span element).

Approving PR in advance
